### PR TITLE
fix(statedb): Correctly implement StateDB's deep copy

### PR DESCRIPTION
**Pull request by calbera**
- Correct statedb behavior in accordance to geth 
- Simplify Polaris interfaces where possible

### CODE REVIEW

**Comment by calbera** [comment]
getting the build error: "import cycle not allowed in test". any idea why?
```
ERROR Running error: context loading failed: failed to load packages: 
failed to load with go/packages: internal error: go list gives conflicting information for package 
pkg.berachain.dev/polaris/eth/core/state/journal [pkg.berachain.dev/polaris/eth/core/state/journal.test]
```

File: eth/core/processor.go [hunk 1]
```diff
@@ -239,6 +239,7 @@ func (sp *StateProcessor) BuildAndRegisterPrecompiles(precompiles []precompile.R
 		if err != nil {
 			panic(err)
 		}
+		// TODO: set code on the statedb for every precompiled contract.
```
Comment (itsdevbear) [inline eth/core/processor.go:242]: nice

File: eth/core/state/interfaces.go [hunk 2]
```diff
@@ -38,6 +37,8 @@ type Plugin interface {
 	libtypes.Preparable
 	// Reset resets the state with the given `context`.
 	libtypes.Resettable
+	// Plugin implements `libtypes.Cloneable`.
```
Comment (itsdevbear) [inline eth/core/state/interfaces.go:40]: cloneable good proj

File: eth/core/state/journal/logs.go [hunk 3]
```diff
@@ -25,21 +25,38 @@ import (
 	coretypes "pkg.berachain.dev/polaris/eth/core/types"
 	"pkg.berachain.dev/polaris/lib/ds"
 	"pkg.berachain.dev/polaris/lib/ds/stack"
+	libtypes "pkg.berachain.dev/polaris/lib/types"
+	"pkg.berachain.dev/polaris/lib/utils"
 )
 
+// LogsI defines the interface for tracking logs created during a state transition.
+type LogsI interface {
```
Comment (itsdevbear) [inline eth/core/state/journal/logs.go:33]: using the `I` terminology is frowned upon. StateDBLogs?

File: eth/core/state/journal/refund.go [hunk 4]
```diff
@@ -79,3 +92,15 @@ func (r *refund) RevertToSnapshot(id int) {
 func (r *refund) Finalize() {
 	r.Stack = stack.New[uint64](initCapacity)
 }
+
+// Clone implements `libtypes.Cloneable`.
+func (r *refund) Clone() RefundI {
```
Comment (itsdevbear) [inline eth/core/state/journal/refund.go:97]: `I` notatation everywhere

File: eth/core/state/journal/suicides.go [hunk 5]
```diff
@@ -43,6 +45,19 @@ type suicideStatePlugin interface {
 	SubBalance(common.Address, *big.Int)
 }
 
+type SuicidesI interface {
```
Comment (itsdevbear) [inline eth/core/state/journal/suicides.go:48]: +1

**Review by itsdevbear — COMMENTED** [review]
Overall makes sense to me, though you removed a bunch of tests lol.

I think we should just make the Controller Cloneable, thoughts?

File: cosmos/x/evm/plugins/precompile/plugin.go [hunk 6]
```diff
@@ -164,7 +163,14 @@ func (p *plugin) Run(
 // EnableReentrancy sets the state so that execution can enter the EVM again.
 //
 // EnableReentrancy implements core.PrecompilePlugin.
-func (p *plugin) EnableReentrancy(_ context.Context) {
+func (p *plugin) EnableReentrancy(evm ethprecompile.EVM) {
+	sdb := utils.MustGetAs[vm.PolarisStateDB](evm.GetStateDB())
```
Comment (itsdevbear) [inline cosmos/x/evm/plugins/precompile/plugin.go:167]: Can we move the re-entrancy guard to the `eth` folder?

Comment (calbera) [reply cosmos/x/evm/plugins/precompile/plugin.go:167]: this is logic specific to Cosmos

Comment (calbera) [reply cosmos/x/evm/plugins/precompile/plugin.go:167]: that we need to stop/start emitting logs, switch gas config

Comment (itsdevbear) [reply cosmos/x/evm/plugins/precompile/plugin.go:167]: Check slack my bad

**Comment by calbera** [comment]
didnt remove tests, and yeah I was thinking Controller could be cloneable but massive lift, too much for now

**Comment by itsdevbear** [comment]
I guess controller you end up in generic hell

**Comment by calbera** [comment]
I can add a couple more unit tests though

**Comment by itsdevbear** [comment]
Add some tests that ensure that if you modify the clones it doesn't modify the original 🫡

**Comment by calbera** [comment]
should be good to go, the pending nonce integration test thats failing is unaffected by this change, which btw it passes for me locally

**Comment by calbera** [comment]
Just verified in geth that `.Copy()` is only called on the StateDB for read-only things like traces and is not committed.
image

**Review by itsdevbear — APPROVED** [review]

**Comment by itsdevbear** [comment]
we can cleanup later

**Pull request merged**